### DISCUSSION TITLE
Prisma Nexus Apollo jwtAuthentication on Angular - am I doing it right?

### DISCUSSION THREAD

**Discussion by TomasMoskala** [prisma/prisma#2780] [Q&A] [answered]
Hi, 
it was quite a struggle to setup everything so it works and my code is different from all I've seen so I am not 100% sure about it. Especially that I needed to declare another 'user' type as a simpler tructure just for the token and the magic behind 'req'  becoming 'ctx'. I also get a little confused about the 2 contexts in server instance and in makeSchema.

server.ts
```
require('dotenv').config()
import { ApolloServer} from 'apollo-server'
import { schema } from './schema'
import { createContext } from './context'
const port = process.env.PORT;

new ApolloServer({
  schema,
  context: ({ req }) => createContext(req),
}).listen(
  { port: port },
  () =>
    console.log(
      `🚀 Server ready at: http://localhost:${port}\n⭐️ See sample queries: http://pris.ly/e/ts/graphql-apollo-server#using-the-graphql-api`,
    ),
)

```
schema.ts
```
import { nexusPrismaPlugin } from 'nexus-prisma'
import { makeSchema } from '@nexus/schema'
import * as types from './types'

export const schema = makeSchema({
  types,
  plugins: [nexusPrismaPlugin()],
  outputs: {
    schema: __dirname + '/../schema.graphql',
    typegen: __dirname + '/generated/nexus.ts',
  },
  typegenAutoConfig: {
    contextType: 'Context.Context',
    sources: [
      {
        source: '@prisma/client',
        alias: 'prisma',
      },
      {
        source: require.resolve('./context'),
        alias: 'Context',
      },
    ],
  },
})

```
context.ts
```
import jwt from 'jsonwebtoken'
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

type user = {
  user_email: string
  user_id: number
}

export interface Context {
  prisma: PrismaClient,
  user: user
}

const getUser = (token: string) => {
  try {
    if (token) {
      return jwt.verify(token, process.env.JWT_SECRET || 'any') as user
    }
    return null
  } catch (err) {
    return null
  }
}

export function createContext(req: any): Context {
    const tokenWithBearer = req.headers.authorization || ''
    const token = tokenWithBearer.split(' ')[1]
    let user = getUser(token)
    if (!user) 
      user = { user_email: '', user_id: 0 }
    return {
      user,
      prisma, 
    }
}

```
query.ts (I have not handled the password yet).

```
export const Query = queryType({
  definition(t) {
 t.field('login', {
      type: AuthPayload,
      args: {
        user_email: stringArg({ nullable: false }),
        password: stringArg({ nullable: false }),
      },
      resolve: async(_, { user_email, password } , ctx) => { 
        const user = await ctx.prisma.user.findOne({
          where: { user_email },   
        });
        if (user === null) {
          throw new Error(`No user found for email: ${user_email}`)
        }
        const passwordValid = password // await compare(password, user.password)
        if (!passwordValid) {
          throw new Error('Invalid password')
        }
        return {
          user,
          token: sign({ user_email: user.user_email, user_id: user.user_id }, process.env.JWT_SECRET || 'JWT_SECRET'),
        }
      },
    })

    t.field('me', {
      type: 'user',
      nullable: true,
      resolve: (_, __, ctx) => {
        const user_email = ctx.user.user_email
        return ctx.prisma.user.findOne({
          where: { user_email },   
        });
      },
    })
}

```

authPayLoad.ts

```
import { objectType } from '@nexus/schema'
import { User } from './user'

export const AuthPayload = objectType({
  name: 'AuthPayload',
  definition(t) {
    t.string('token')
    t.field('user', { type: User })
  },
})

```
types.ts exports in a usual way
```
export const User = objectType({
  name: 'user',
  definition(t) {
    t.model.user_email()
    t.model.user_id()
    t.model.user_password()
etc...
}})

```
It works fine but is it correct??
Thanks

PS: Why does everybody put login into mutations when it does not change anything?

**Comment by ryands17** [comment, marked as answer]
Hey @TomasMoskala :wave: 
Yes this approach is correct and it is in our examples [here](https://github.com/prisma/prisma-examples/tree/master/typescript/graphql-auth) as well.

> PS: Why does everybody put login into mutations when it does not change anything?

Login does create a token for the specific user and that could be considered as an addition, especially if you're managing multiple tokens or a session based authentication. Which is why login is usually kept in mutations.

> types.ts exports in a usual way
> ```ts
> export const User = objectType({
>   name: 'user',
>   definition(t) {
>     t.model.user_email()
>     t.model.user_id()
>     t.model.user_password()
> etc...
> }})
> ```

Also I would suggest not exposing the password field to the client API.

**Reply by TomasMoskala** [reply]
Thanks. The Apollo server is quite a lot different. Now I struggle with the client.

**Comment by Sytten** [comment]
I might just suggest my new plugin if you need help to build an authorization layer: https://github.com/Sytten/nexus-shield
It is young, but we use it in prod at company.

**Comment by ludralph** [comment]
Hi @TomasMoskala,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
